### PR TITLE
Milestone 2: Enumerable Method - Microverse Project

### PR DESCRIPTION
**Pull request by happiguru**
### Enumerable Methods

This pull request contains a project implementing Enumerable Methods developed with Ruby.
We have implemented the following methods:

1. - my_each method
2. - my_each_with_index method
3. - my_select method
4. - my_all? method
5. - my_any? method
6. - my_none? method
7. - my_count method
8. - my_map method
9. - my_inject method
10. - multiply_els method

_Quite interesting to develop our own method....._

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    return to_enum(:my_each) unless block_given?
+
+    arr = self.class == Array ? self : to_a
+    counter = 0
+    while counter < arr.length
+      yield(arr[counter])
+      counter += 1
+    end
+    arr
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Range and array self
- [ ] Test with hashes and ranges, then compare with original `each` Ruby method
- [ ] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_each{
# or
Range.new(0, 9)

```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    return to_enum(:my_each_with_index) unless block_given?
+
+    arr = self.class == Array ? self : to_a
+    counter = 0
+    while counter < arr.length
+      yield(arr[counter], counter)
+      counter += 1
+    end
+    arr
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Range and array self
- [ ] Test with hashes and ranges, then compare with original `each_with_index` Ruby method
- [ ] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_each_with_index{
# or
Range.new(0, 9)

```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+  # My Select
+  def my_select(arr)
+    new_array = []
+    arr.my_each do |index_value|
+      new_array << index_value if yield(index_value)
+    end
+    new_array
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Enumerator error
- [ ] It should raise a "Enumerator" error when no block or argument is given

## Range and array self
- [ ] Test with hashes and ranges, then compare with original `select` Ruby method
- [ ] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_select{
# or
Range.new(0, 9)


```

## Optional argument or none argument
- [x] Add optional argument or remove to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_select?(argument = nil)
def my_select?
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+  # My All?
+  def my_all?(arr)
+    result = false
+    arr.my_each do |item|
+      result = true if yield(item == true)
+    end
+    result
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Optional argument
- [ ] Add  optional argument to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_all?(argument = nil)
```

## Hashes, range and array self
- [ ] Test with hashes and ranges, then compare with original `all` Ruby method
- [ ] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_all{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  # My Any?
+  def my_any?(arr)
+    result = false
+    arr.my_each do |item|
+      next unless yield(item)
+
+      result = true
+    end
+    result
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Optional argument
- [ ] Add  optional argument to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_any?(argument = nil)
```

## Hashes, range and array self
- [x] Test with hashes and ranges, then compare with original `any` Ruby method
- [x] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_any{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+  # My None?
+  def my_none?(arr)
+    result = true
+    arr.my_each do |item|
+      result = false if yield(item)
+    end
+    result
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Optional argument
- [ ] Add  optional argument to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_none?(argument = nil)
```

## Hashes, range and array self
- [ ] Test with hashes and ranges, then compare with original `none` Ruby method
- [ ] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_none{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+  # My Count
+  def my_count(arr)
+    counter = 0
+    arr.my_each do |item|
+      counter += 1 if yield(item)
+    end
+    counter
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Optional argument
- [ ] Add  optional argument to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_none?(argument = nil)
```

## Hashes, range and array self
- [x] Test with hashes and ranges, then compare with original `count` Ruby method
- [x] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_count{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+  # My Map
+  def my_map(arr, &block_given)
+    result = []
+    arr.my_each do |element|
+      result << block_given.call(element)
+    end
+    result
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## Optional argument
- [ ] Add  optional argument to this method. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)


```ruby
def my_map?(argument = nil)
```

## Hashes, range and array self
- [x] Test with hashes and ranges, then compare with original `map` Ruby method
- [x] Fix your method to work with the tests. Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

**Example:**

`test_file.rb`
```ruby
(0..9).my_map{
# or
Range.new(0, 9)

# HASHES
{ a: 1, b: 2, c: 3, d: 4, e: 5 }
```

## Enumerator error
- [x] It should raise an "Enumerator" error when no block or argument is given

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  # My Inject
+  def my_inject(arr, start_value = 0)
+    counter = 0
+    accumulator = start_value
+    while counter < arr.size
+      accumulator = yield(accumulator, arr[counter])
+      counter += 1
+    end
+    accumulator
+  end
```
Comment (bertil291utn) [inline enumerables.rb]: ## LocalJumpError
- [ ] It should raise a "LocalJumpError" error when no block or argument is given

## Zero or many arguments
- [ ] Add  Zero or many arguments to this method

```ruby
def my_inject?(*argument)
```

**Note:** Remember this method allows procs, block, no block. Thus check the original `inject` ruby function or try to follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+  # Multiply LS
+  def multiply_els(arr)
+    arr.my_inject(arr, 1) { |a, b| a * b }
+  end
+end
```
Comment (bertil291utn) [inline enumerables.rb]: Follow this [sample](https://gist.github.com/bertil291utn/5a7e6a73351c38ceb9781a9a37b58cf6)

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+puts 'my-count'
+puts my_arr.my_count(my_arr) { |n| (n % 2).zero? }
+
+puts 'my_map'
+puts my_arr.my_map(my_arr) { |n| n + 4 }
+puts 'my_inject'
+puts my_arr.my_inject(my_arr, 3) { |n, y| n + y }
+
+puts 'multiply_els'
+puts my_arr.multiply_els(my_arr)
```
Comment (bertil291utn) [inline enumerables.rb]: ## Debugging instructions
- No debugging instructions in committed code (as discussed [here](https://questions.microverse.org/t/debugging-instructions-code/346)) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/section/debugging_requirement.png) from Pathwright)
  - [x] Please remove them

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish all requirements, you required to do these changes, listed below:
- Check all the inline comments above this box
### OPTIONAL
- [ ] It'd be great if you add the number one section below the screenshot. In the top, you describe what the app does, below add another section to count all method you have

### Summary
- Try to change all your arguments ones should be optional another one should be just zero or many arguments.
- Check with range and self arrays `[0,1,2].my_each`, range `(0..9).my_each` and Hashes `{1:'Peter'}.my_each`
-  Don't forget to work with inline comments below this comment box.

***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

File: enumerables.rb (hunk 1 above)
Comment (happiguru) [reply enumerables.rb]: Changed the structure of the code tested the code to ensure the above test examples are working perfectly.

File: enumerables.rb (hunk 2 above)
Comment (happiguru) [reply enumerables.rb]: Removed the while loop and made sure it passes all the required test.

File: enumerables.rb (hunk 1 above)
Comment (happiguru) [reply enumerables.rb]: Refactor the code: Removed the outdated push symbols and replaced it with push built-in method.

Comment (happiguru) [reply enumerables.rb]: All test were tested and working properly

File: enumerables.rb (hunk 4 above)
Comment (happiguru) [reply enumerables.rb]: Refactor Code: Added test conditions with the use of switch statements.
We have also tested the code for various datatypes and it works fine

File: enumerables.rb (hunk 5 above)
Comment (happiguru) [reply enumerables.rb]: Updated the code to adjust it for several data inputs.

File: enumerables.rb (hunk 8 above)
Comment (happiguru) [reply enumerables.rb]: Fixed and tested

File: enumerables.rb (hunk 9 above)
Comment (happiguru) [reply enumerables.rb]: Fixed and tested

File: enumerables.rb (hunk 10 above)
Comment (happiguru) [reply enumerables.rb]: Fixed and tested

File: enumerables.rb (hunk 11 above)
Comment (happiguru) [reply enumerables.rb]: Fixed and tested

File: enumerables.rb (hunk 7 above)
Comment (happiguru) [reply enumerables.rb]: Fixed and tested

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+      if !block_given?
+        case argument
+        when argument.nil?
+          next if item
+        when argument.class == Class
+          next if item.is_a? argument
+        when argument.class == regexp
+          next if item =~ argument
+        when (item.is_a? Numeric) || (item.is_a? String)
+          next if item == argument
```
Comment (expjazz) [inline enumerables.rb]: - [ ] There is a bug here. This:
_[1, 'a', true].all?_ should return true but is throwing an error instead. The 'regex' needs to be 'Regex'.

- [ ] Even if you fix that problem, this is returning false when it should be true for all the arguments (Class, regexp and etc.).

image

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    Array(self).my_each do |arg|
+      if !block_given?
+        case arr
+        when arr.nil?
+          next unless arg
+        when arr.class == Class
+          next unless arg.is_a? arr
+        when arr.class == Regexp
+          next unless arg =~ arr
+        when (arg.is_a? Numeric) || (arg.is_a? String)
```
Comment (expjazz) [inline enumerables.rb]: - [ ] Same Problem of the my_all. You need to fix this for all the possible arguments.

image

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+          next unless arg == arr
+        end
+      else
+        next unless yield(arg)
+      end
+      return false
+    end
```
Comment (expjazz) [inline enumerables.rb]: - [ ] Same problem of the previous methods.

image

File: enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        return Array(self).length if arg.nil?
+        next if arg != arr
+
+        counter += 1
+      elsif yield(arg) == true
+        counter += 1
+      end
+    end
+    counter
+  end
```
Comment (expjazz) [inline enumerables.rb]: - [x] It needs to return the count of items in the enum.

image

File: enumerables.rb [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    Array(self).my_each_with_index do |item, index|
+      next if skip_flag && index.zero?
+
+      if block_given?
+        acummulator = yield(acum, item)
+      elsif arguments[0].class == Symbol
+        acummulator = acummulator.send(arguments[0], item)
+      elsif arguments[0].is_a? Numeric
+        acummulator = acummulator.send(arguments[1], item)
+      end
```
Comment (expjazz) [inline enumerables.rb]: - [ ]It is throwing an error. First, fix the acum error in line 141 . Also, uncomment line 128

image

**Review by expjazz — CHANGES_REQUESTED** [review]
## Status: Changes Required :recycle: 

Congrats! Your project is awesome and you are really close :clap: 
You got a few errors in typing in your code, and a few tweaks to do. Kindly check the inline comments for more details.

### Please make the required changes and resubmit for a new code review once you are ready
Good Luck and Happy Coding :pray:

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Expedito Andrade).:smile:_

File: enumerables.rb (hunk 16 above)
Comment (happiguru) [reply enumerables.rb]: Updated: Fix the bug and did uncomment line 128

File: enumerables.rb (hunk 15 above)
Comment (happiguru) [reply enumerables.rb]: Fixed: my_count method has been updated to ensure it returns a count of the items in the array

File: enumerables.rb (hunk 14 above)
Comment (happiguru) [reply enumerables.rb]: Fixed: Refactor code to yield the expected result

File: enumerables.rb (hunk 13 above)
Comment (happiguru) [reply enumerables.rb]: Issue fixed

File: enumerables.rb (hunk 12 above)
Comment (happiguru) [reply enumerables.rb]: Issue Fixed

**Review by addod19 — APPROVED** [review]
## Status: Approved :trophy: :balloon: 

Hi **TEAM**, congratulations on fixing all the comments made by the previous TSE's :100: Our project now meets all the minimum requirements, I hereby approve of your project :fireworks: 
Good luck with your next projects

Go ahead and merge into development :slightly_smiling_face: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Pull request merged**